packages/gatsby/src/redux/actions/public.js

```diff
@@ -47,6 +48,19 @@ const ensureWindowsDriveIsUppercase = filePath => {
     : filePath
 }
 
+const truncatePath = path => {
+  if (path === `/`) return path
+  const pathParts = path.split(`/`).filter(Boolean)
+  const truncatedParts = pathParts.map(pathPart => {
+    // OS limit is 255 characters
+    if (pathPart.length > 255) {
+      return pathPart.slice(0, 100) + strhash(pathPart.slice(100))
```

Comment:
we already have some hash creation util [`createContentDigest` from `gatsby-core-utils`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-core-utils#createcontentdigest) - it will create different hash than `string-hash`, but it should still fit as it will be 32 alphanumerical characters